File: Problems/Algorithms/547. Number of Provinces/NumberOfProvinces.kt
class Solution {
    fun findCircleNum(isConnected: Array<IntArray>?): Int {
        if (isConnected == null || isConnected.size == 0) {
            return 0
        }
        
        val n = isConnected.size
        val uf = UnionFind(n)
        for (i in 0..n-1) {
            for (j in 0..n-1) {
                if (isConnected[i][j] == 1) {
                    uf.union(i, j)
                }
            }
        }
        
        return uf.getCount()
    }
    
    private class UnionFind(size: Int) {
        private val root = IntArray(size) {i -> i}
        private val rank = IntArray(size) {i -> 1}
        private var count = size
        
        fun find(x: Int): Int {
            if (x == root[x]) {
                return x
            }
            root[x] = find(root[x])
            return root[x]
        }
        
        fun union(x: Int, y: Int): Unit {
            val rootX = find(x)
            val rootY = find(y)
            
            if (rootX != rootY) {
                if (rank[rootX] > rank[rootY]) {
                    root[rootY] = rootX
                } else if (rank[rootX] < rank[rootY]) {
                    root[rootX] = rootY
                } else {
                    root[rootY] = rootX
                    rank[rootX] += 1
                }
                count--
            }   
        }
        
        fun getCount(): Int {
            return count
        }
    }
}

File: Problems/Algorithms/547. Number of Provinces/number-of-provinces.cpp
#include <vector>

using namespace std;

class Solution {
public:
    void dfs(int node, vector<vector<int>>& isConnected, vector<bool>& visited) {
        visited[node] = true;
        for (int i = 0; i < isConnected.size(); i++) {
            if (isConnected[node][i] && !visited[i]) {
                dfs(i, isConnected, visited);
            }
        }
    }

    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        vector<bool> visited(n);

        int ans = 0;
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                ans++;
                dfs(i, isConnected, visited);
            }
        }

        return ans;
    }
};
